lib/ansible/plugins/action/gather_facts.py

```diff
@@ -41,7 +41,16 @@ def _get_module_args(self, fact_module, task_vars):
         mod_args = dict((k, v) for k, v in mod_args.items() if v is not None)
 
         # handle module defaults
-        mod_args = get_action_args_with_defaults(fact_module, mod_args, self._task.module_defaults, self._templar, self._task._ansible_internal_redirect_list)
+        if fact_module == 'ansible.legacy.setup' and self._use_setup_defaults:
```

Comment:
I think probably all the changes to the actions can go away if you were to change get_action_args_with_defaults to say:
* if `action` is prefixed with `ansible.legacy.` and the same `ansible.legacy`-prefixed action appears in the redirect list, remove the prefix from both places before looping over to apply module_defautls